fish-rust/src/builtins/string/match.rs

```diff
@@ -0,0 +1,409 @@
+            'n' => self.index = true,
+            _ => return Err(StringError::UnknownOption),
+        }
+        return Ok(());
+    }
+
+    fn take_args(
+        &mut self,
+        optind: &mut usize,
+        args: &[&'args wstr],
+        streams: &mut io_streams_t,
+    ) -> Option<libc::c_int> {
+        let cmd = args[0];
+        let Some(arg) = args.get(*optind).copied() else {
+               string_error!(streams, BUILTIN_ERR_ARG_COUNT0, cmd);
+               return STATUS_INVALID_ARGS;
+           };
+        *optind += 1;
+        self.pattern = arg;
+        STATUS_CMD_OK
+    }
+
+    fn handle(
+        &mut self,
+        parser: &mut parser_t,
+        streams: &mut io_streams_t,
+        optind: &mut usize,
+        args: &[&wstr],
+    ) -> Option<libc::c_int> {
+        let cmd = args[0];
+
+        if self.entire && self.index {
+            streams.err.append(wgettext_fmt!(
+                BUILTIN_ERR_COMBO2,
+                cmd,
+                "--entire and --index are mutually exclusive"
+            ));
+            return STATUS_INVALID_ARGS;
+        }
+
+        if self.invert_match && self.groups_only {
+            streams.err.append(wgettext_fmt!(
+                BUILTIN_ERR_COMBO2,
+                cmd,
+                "--invert and --groups-only are mutually exclusive"
+            ));
+            return STATUS_INVALID_ARGS;
+        }
+
+        if self.entire && self.groups_only {
+            streams.err.append(wgettext_fmt!(
+                BUILTIN_ERR_COMBO2,
+                cmd,
+                "--entire and --groups-only are mutually exclusive"
+            ));
+            return STATUS_INVALID_ARGS;
+        }
+
+        let mut matcher = match StringMatcher::new(self.pattern, self) {
+            Ok(m) => m,
+            Err(e) => {
+                e.print_error(args, streams);
+                return STATUS_INVALID_ARGS;
+            }
+        };
+
+        for (arg, _) in Arguments::new(args, optind, streams, true) {
+            if let Err(e) = matcher.report_matches(arg.as_ref(), streams) {
+                FLOG!(error, "pcre2_match unexpected error:", e.error_message())
+            }
+            if self.quiet && matcher.match_count() > 0 {
+                break;
+            }
+        }
+
+        let match_count = matcher.match_count();
+
+        if let StringMatcher::Regex(RegexMatcher {
+            first_match_captures,
+            ..
+        }) = matcher
+        {
+            let vars = parser.get_vars();
+            for (name, vals) in first_match_captures.into_iter() {
+                vars.set(&WString::from(name), EnvMode::DEFAULT, vals);
+            }
+        }
+
+        if match_count > 0 {
+            STATUS_CMD_OK
+        } else {
+            STATUS_CMD_ERROR
+        }
+    }
+}
+
+struct RegexMatcher<'opts, 'args> {
+    regex: Box<Regex>,
+    total_matched: usize,
+    first_match_captures: HashMap<String, Vec<WString>>,
```

Comment:
Since I was curious: the capture-groups names is an UTF8-string from [here](https://github.com/fish-shell/rust-pcre2/blob/master/src/ffi.rs#L68). This is currently handled by [this](https://github.com/fish-shell/fish-shell/blob/63b23713f22f0d006c8bfb011c1c2b621d25cb73/src/re.cpp#L196) in `re.cpp`. When `pcre2` is not in UTF-mode, the capture group name is limited to alphanumerics and underscores according to the [docs](https://www.pcre.org/current/doc/html/pcre2syntax.html#SEC13). I just learned that you can [enable UTF-mode](https://www.pcre.org/current/doc/html/pcre2unicode.html) with `(*UTF)` so `string match -r "(*UTF)(?<aΩ>.*)" "aaa"`. Running that normally lead to

```
string match: Regular expression compile error: using UTF is disabled by the application
string match: (*UTF)(?<aΩ>.*)
string match:
```

But we have not set `never UTF`, will fix

Reply:
Now rebased, I will mark this as resolved